framework: error handling and response for REST (#35)
inveniosoftware/rfcs: rfcs/framework-0035-error-handling-and-response.md

At a client-level, the most important is having a machine-readable format for the error, and especially it's important for Single-Page applications in order to display errors. The response will always have a content type defined, and thus it's reasonable striaghtforward to handle errors in a different format that what was requested, especially if it is only for errors.

**Error response without content negotiation (example 2)**

Having only a single response format for errors simplifies all of above mentioned issues with the content negotiation. Serializers needs only to deal with how to serialize a successful response and not a pletora of errors, and thus becomes easier to write. Having only one format, means errors can easily be generated from multiple different sources and produce the same error (e.g. 404 from service vs routing layer), and clients can easily decode errors and know it's always the same format.

> Because of above reasons, it means that *error response bodies are not subject to content negotiation and thus always JSON formatted*

Author: You have to separate the two cases 1) developing Invenio Framework and 2) using Invenio Framework to build applications.

This RFC only deals with (1). It's not supposed to be prescriptive about (2). If a person using Invenio wants to make a ContentNegotiatingException that's able to reply in XML and JSON it's fine.

I'll try to update the intro to make this clear.

> Because of above reasons, it means that *error response bodies are not subject to content negotiation and thus always JSON formatted*

Reviewer: Right. I think this is sort of dealt with by the possibility of creating custom exceptions that know how to render themselves potentially.

### Who is responsible for rendering the HTTP response body for an error?

Having established that error response are not subject to content negotiation, the next question is who is responsible for rendering the HTTP response body for an error.

As a prerequisite, please see [error handling in Flask](https://flask.palletsprojects.com/en/1.1.x/errorhandling/#error-handlers).

**Sources of errors**

The following are usual sources of errors (i.e. exceptions being raised):

1. Flask layer, for instance routing errors.
1. View layer, for instance ``abort(400)`` or ``raise BadRequest()``.
1. Service/data source layer, for instance Elasticsearch ``RequestError`` or ``PIDDeleteError``.

> A bit oversimplified, you can say that Flask/view layer errors usaully  result in HTTP errors such as 40x errors (Bad Request, Not Found).

Reviewer 2: ```suggestion
A bit oversimplified, you can say that Flask/view layer errors usually result in HTTP errors such as 40x errors (Bad Request, Not Found).
```

Service/data source layer errors will usually result in 500 Internal Server Error *unless they are handled*.

For all above type of errors, we'd like to create a machine readable error response.

**View error handling**

A Flask view can simply catch exceptions, and render a response. This can either be done per view or via special decorators.

```python
@app.route('/')
def view1():
    try:
        # ...
    except Exception:
        return 400, ...
```

The view layer error handling is best suited for very custom error handling from the service layer. E.g implementation of a custom API with custom reposnses.

The issue with view layer error handling is that only errors from the view and service/data source layer is being caught. Any error in the Flask layer (routing, request processing functions etc) will escape the try/except-clause (because they happen outside the function).

Example of a 404 error:

1. A routing error will result in a 404 error from the Flask layer.
2. An object that doesn't exists in the database, will likely result in a 404 error from the service layer.

> A view layer can only catch the latter. Thus, if we render an error response in the view, we still have to deal with error response for Flask layer errors.

Author: Where would the 400 error come from in the Flask-layer? Query parsing is a view layer error?

**Flask's error handling/rendering**

Flask and Werkzeug's request dispatching system will catch all errors happening in the three layers described. The error handling is pretty sophisticated, and is explained below.

First, Flask will look for a user-provided error handler. If found, the handler is responsible for making a response. If not found, but the exception is an HTTPException (or subclass thereof), the HTTPException will simply be returned, because an HTTPException can self-render. If it's not an HTTPException, the error is simply reraised and results in an Internal Server Error. The following pseudo-code is an example of the handling:

```python

> def handle_exception(e):
>     # If it's an HTTPException, use different method:
>     if isinstance(e, HTTPException):
>         return handle_http_exception(e)
> 
>     handler = find_error_handler(e)
>     if handler is None:
>         # No handler was found, and it's not and HTTPException.
>         raise # <----- reraise
>     # A handler was found, use it to return response
>     return handler(e)
> 
> def handle_http_exception(e):
>     handler = find_error_handler(e)
>     if handler is None:
>         # No handler was found, but it is an HTTPException.
>         return e # <----- return
>     # A handler was found, use it to return response
>     return handler(e)

Reviewer: Not sure I follow why HttpExceptions are treated differently. The pseudo-code above simplifies to:

 ```suggestion
def handle_exception(e):
    handler = find_error_handler(e)
    if handler is None:
        # No handler was found
        raise e  # <----- reraise. This results in a 500 if e is not an HTTPException or 
                 # the appropriate status code because it is an HTTPException
    # A handler was found, use it to return response
    raise handler(e)

```

Author: It's because the "pseudo-code" is taken directly from Flask were there are two methods, so I prefer to keep it as close as possible to Flask code.

```

The key in understanding the difference between an normal exception and an HTTPException is that an HTTPException is considered a valid response by Flask's ``make_reponse``. This allows it to be returned by the ``handle_http_exception()`` function.

**Recap**

Source of errors (exceptions raised) is three layers (Flask, View, Service). For handling these errors (i.e. making an HTTP error response) we can use Flask and Werkzeug's error handling system. The view layer is not well-suited for this.

> 

Reviewer: ∑(O_O;) I think I am not quite convinced that the view layer is not well-suited... Actually, what is puzzling me is that we seem to try to build up the argument for why to use Flask's `error_handler` functionality, but handling errors via a `try-catch` in the decorators of flask-resources or via `error_handler` of Flask (functionality in flask-resources as well) is actually equivalent (or at least the difference hasn't been made clear to me... :sweat_smile: ). In both cases we are placed at the blueprint level...

Author: It's not equivalent because Flask-level errors escape the view layer.

Author: What this means is basically you'll need both Flask errors handlers and you view error. Instead, you can simply let the Flask error handler do the entire job and not duplicate error handling.

Above basically establish that either

1. an *exception* can render itself into an HTTP error response, or
2. an *error handler* can render an exception into a HTTP error response.

### Error handling in Flask REST API applications

Following is a proposal for how to implement the error handling in REST API application

**JSON-formatted HTTP Exception**

First we propose to create a new subclass of ``HTTPException``, that will render the response body as JSON. In simplified pseudo-Python  this would be something like:

```python
class HTTPJSONException(HTTPException):
    def get_body(self, environ=None):
        body = dict(
            status=self.code,
            message=self.get_description(environ=environ),
        )
        return json.dumps(body)):
```

**Flask application**

Next, we install an error handler on the Flask application for ``HTTPExceptions``:

```python
from Flask import Flask

app = Flask(__name__)

> @app.errorhandler(HTTPException)

Reviewer: I think that's the core of my disagreement: in invenio-records-resources this can be done at the app level but in flask-resources this should be done on a blueprint **only**.

Author: The RFC is about error handling principles in an Invenio-based application, and it is not about if Flask-Resources or Invenio-Records-Resources should install application-level error handlers. The main purpose is to document the thoughts behind the *current* approach.

Already today, Invenio-REST is installing the application-level error handlers, and Invenio-Records-REST is installing blueprint level error handlers more or less according to what is described here. There are few details in the current implementation that could be simplified and improved, but overall the approach is the documented one. So you can say this RFC is simply confirming the current approach. 

I think it's probably more a misunderstanding of scope that needs to better documented in the RFC.

Reviewer: Probably the scope is throwing me off here :sweat_smile: . Is this how we want error handling to be managed in the future (in flask-resources and invenio-records-resources) or is it how it *is* working in production invenio (invenio-rest)? Maybe the RFC cuts things differently and that is not even a concern... :thinking: ?

- Start Date: 2020-07-15
- RFC PR: [#35](https://github.com/inveniosoftware/rfcs/pull/35)

> - Authors: Lars Holm Nielsen, Zach Zacharodimos

Reviewer 3: If you want my official name is `Zacharias Zacharodimos` hehe :)

Because of above reasons, it means that *error response bodies are not subject to content negotiation and thus always JSON formatted*

### Who is responsible for rendering the HTTP response body for an error?

Having established that error response are not subject to content negotiation, the next question is who is responsible for rendering the HTTP response body for an error.

IMPORTANT: As a prerequisite, please read [error handling in Flask](https://flask.palletsprojects.com/en/1.1.x/errorhandling/#error-handlers).

**Sources of errors**

The following are usual sources of errors (i.e. exceptions being raised):

1. Flask layer, for instance routing errors.
1. View layer, for instance ``abort(400)`` or ``raise BadRequest()``.

> 1. Service/data source layer, for instance Elasticsearch ``RequestError`` or ``PIDDeleteError``.

Reviewer 3: Is it only me or here there 3 `1)`? @lnielsen

```

**Flask application**

Next, we install an error handler on the Flask application for ``HTTPExceptions``:

```python
from Flask import Flask

app = Flask(__name__)

@app.errorhandler(HTTPException)
def handle_http_exception(e):
    return HTTPJSONException(code=e.code, description=e.description)
```

This will ensure that all errors generated at the Flask layer will have a proper JSON-formatted error response.

**Blueprints**

Last, for blueprints containing the views of the REST API, we propose to add error handlers that able to map service errors to ``HTTPJSONException``:

```python
# A function that creates an error handler.
# The error handler, maps the exception into an HTTPException.
def create_error_mapper(exc):
    def handler(e):
        return HTTPJSONException(code=e.code, description=e.description)
    return handler

# A list of service errors, and their handlers
error_map = {
    ServiceError: create_error_mapper(HTTPJSONException),
    # ...
}

# Install handlers on blueprints.
def register_error_map(blueprint, error_map):
    for exc_or_code, handler in error_map.items():
        blueprint.register_error_handler(exc_or_code, handler)
```

#### Usage

> The above error handler technique, can the be used with the existing exceptions and Flask idoms:

Reviewer 2: ```suggestion
The above error handler technique, can the be used with the existing exceptions and Flask idioms:
```
